### PR TITLE
[IMP] product: find best price for a vendor based on min qty

### PR DESCRIPTION
**Pull request by hch-odoo**
task - https://www.odoo.com/web?#id=1947351&action=327&model=project.task&view_type=form&menu_id=4720

pad - https://pad.odoo.com/p/r.42f76c4168c856b53271369fa2674ee7

### CODE REVIEW

File: addons/product/models/product.py [hunk 1]
```diff
@@ -620,9 +620,14 @@ def _select_seller(self, partner_id=False, quantity=0.0, date=None, uom_id=False
             if seller.product_id and seller.product_id != self:
                 continue
 
-            res |= seller
-            break
-        return res
+            res += seller
```
Comment (amoyaux) [inline addons/product/models/product.py]: @hch-odoo remove your diff and .sorted('price') on sellers

Comment (hch-odoo) [reply addons/product/models/product.py]: as per the document https://docs.google.com/document/d/10StevS6AIPnCHUKRmT11GumsifgD9z-FTd6jubpPAj0/edit
we have to consider the sequence of the suppliers first and then we need to check for the ordered quantity.
as you can check in the document, if there are multiple suppliers for the same product but having different price based on ordered quantity, then we have to consider the quantity.

Comment (amoyaux) [reply addons/product/models/product.py]: @hch-odoo then it's an order by quantity desc and sequence but this specification seems strange and hard to understand for classic user, I will check with Laura

Comment (amoyaux) [reply addons/product/models/product.py]: add a condition in this case
``` python
if not seller or seller.name == res.name:
    res |= seller
# after the loop
return res.sorted('price')[0]
```

File: addons/sale_purchase/tests/test_sale_purchase.py [hunk 2]
```diff
@@ -255,3 +255,72 @@ def test_update_ordered_sale_quantity(self):
         self.assertEqual(len(purchase_order2.order_line), 1, "The 2nd PO has only one line")
         self.assertEqual(purchase_line2.sale_line_id, self.sol1_service_purchase_1, "The 2nd PO line came from the SO line sol1_service_purchase_1")
         self.assertEqual(purchase_line2.product_qty, delta, "The quantity of the new PO line is the quantity added on the Sale Line, after first PO confirmation")
+
+    def test_choose_supplier(self):
+        """ Choose supplier based on the ordered quantity and minimum price"""
+        product_to_buy = self.env['product.product'].create({
+            'name': "Furniture Service",
```
Comment (amoyaux) [inline addons/sale_purchase/tests/test_sale_purchase.py]: reduce the args to the required fields and fields necessary for the test -> readability

File: addons/sale_purchase/tests/test_sale_purchase.py [hunk 3]
```diff
@@ -255,3 +255,72 @@ def test_update_ordered_sale_quantity(self):
...
+            'expense_policy': 'no',
+            'default_code': 'SERV_FUR',
+            'service_type': 'manual',
+            'service_to_purchase': True,
+        })
+        supplierinfo1 = self.env['product.supplierinfo'].create({
+            'name': self.partner_vendor_service.id,
+            'price': 100,
+            'product_tmpl_id': self.service_purchase_2.product_tmpl_id.id,
+            'product_tmpl_id': product_to_buy.product_tmpl_id.id,
```
Comment (amoyaux) [inline addons/sale_purchase/tests/test_sale_purchase.py]: you have a duplicated key product_tmpl_id

File: addons/sale_purchase/tests/test_sale_purchase.py [hunk 4]
```diff
@@ -255,3 +255,72 @@ def test_update_ordered_sale_quantity(self):
...
+            'min_qty': 2
+        })
+        supplierinfo2 = self.env['product.supplierinfo'].create({
+            'name': self.partner_vendor_service.id,
+            'price': 60,
+            'product_tmpl_id': self.service_purchase_2.product_tmpl_id.id,
+            'product_tmpl_id': product_to_buy.product_tmpl_id.id,
```
Comment (amoyaux) [inline addons/sale_purchase/tests/test_sale_purchase.py]: dup key

File: addons/sale_purchase/tests/test_sale_purchase.py [hunk 5]
```diff
@@ -255,3 +255,72 @@ def test_update_ordered_sale_quantity(self):
...
+            'min_qty': 10
+        })
+        supplierinfo3 = self.env['product.supplierinfo'].create({
+            'name': self.partner_vendor_service.id,
+            'price': 80,
+            'product_tmpl_id': self.service_purchase_2.product_tmpl_id.id,
+            'product_tmpl_id': product_to_buy.product_tmpl_id.id,
```
Comment (amoyaux) [inline addons/sale_purchase/tests/test_sale_purchase.py]: dup key

File: addons/sale_purchase/tests/test_sale_purchase.py [hunk 6]
```diff
@@ -255,3 +255,72 @@ def test_update_ordered_sale_quantity(self):
...
+            'min_qty': 5
+        })
+        # create sale order having ordered quantity 3.
+        sale_order = self.env['sale.order'].create({
+            'partner_id': self.partner_customer_usd.id,
+            'partner_invoice_id': self.partner_customer_usd.id,
```
Comment (amoyaux) [inline addons/sale_purchase/tests/test_sale_purchase.py]: same than for first comment -> readability

File: addons/product/models/product.py [hunk 7]
```diff
@@ -652,10 +652,9 @@ def _select_seller(self, partner_id=False, quantity=0.0, date=None, uom_id=False
                 continue
             if seller.product_id and seller.product_id != self:
                 continue
-
-            res |= seller
-            break
-        return res
+            if not res or res.name == seller.name:
+                res |= seller
+        return res.sorted('price')[0] if res.ids else res
```
Comment (amoyaux) [inline addons/product/models/product.py]: return res.sorted('price')[:1]

Comment (hch-odoo) [reply addons/product/models/product.py]: done

**Comment by sle-odoo** [comment]
robodoo r+ rebase-ff

**Comment by sle-odoo** [comment]
robooo r+ rebase-ff

**Comment by sle-odoo** [comment]
robooo r+ rebase-ff

**Comment by sle-odoo** [comment]
robodoo r+ robodoo r+ robodoo r+ robodoo r+

**Comment by sle-odoo** [comment]
robodoo retry?

**Comment by sle-odoo** [comment]
robodoo r+ rebase-ff

**Pull request closed without merging**